### DISCUSSION TITLE
Adding guest agent for hypervisor?

### DISCUSSION THREAD

**Discussion by icsy7867** [flatcar/Flatcar#912] [General]
I have read through several docs and I am still a bit confused on this process.  Currently flatcar is running on oVirt and working great.  I would like to add the ovirt guest agent to the running image.  I am confused on how/when/where this can be done.  Would this have to be done before the image is written to disk?  Or can it be done after a successful deployment?

**Comment by tormath1** [comment]
Hi @icsy7867,

I don't know about oVirt but according to the documentation (https://www.ovirt.org/dropped/guest-agent/understanding-guest-agents-and-other-tools.html):
> The oVirt Guest Agent has been deprecated in favor of QEMU Guest Agent

Flatcar ships qemu-guest-agent since release 3402 (https://www.flatcar.org/docs/latest/setup/customization/acpi/#qemu-guest-agent) and you can easily try it:
```
./flatcar_production_qemu.sh -- -chardev socket,path=/tmp/qga.sock,server=on,wait=off,id=qga0 -device virtio-serial -device virtserialport,chardev=qga0,name=org.qemu.guest_agent.0
ssh flatcar
systemctl status qemu*
● qemu-guest-agent.service - QEMU Guest Agent
     Loaded: loaded (/usr/lib/systemd/system/qemu-guest-agent.service; static)
     Active: active (running) since Thu 2022-12-08 14:10:45 UTC; 10s ago
   Main PID: 814 (qemu-ga)
      Tasks: 2 (limit: 30216)
     Memory: 2.1M
        CPU: 18ms
     CGroup: /system.slice/qemu-guest-agent.service
             └─814 /usr/bin/qemu-ga

Dec 08 14:10:45 localhost systemd[1]: Started QEMU Guest Agent.
```

**Reply by icsy7867** [reply]
Thanks for the lightning quick response!

I know this sounds easy.  But i am missing something fundamental.  Where do I find that shell script?

*EDIT*
Found this:
https://flatcar-linux.org/docs/latest/installing/vms/qemu/

But when I try to run the script:
```
qemu-system-x86_64: command not found
```

*EDIT 2*
So it looks like that script is designed to run to launch the VM.  However, oVirt is based on QEMU and KVM, but it has it's own web interface and engine and you cannot simply run a VM on the command line.

**Reply by tormath1** [reply]
Exactly, the `./flatcar_production_qemu.sh` is the fastest way to try Flatcar with QEMU.

From the doc (https://www.flatcar.org/docs/latest/setup/customization/acpi/#qemu-guest-agent):
>  The agent service will automatically be enabled if a virtio-port with the name `org.qemu.guest_agent.0` is detected.

I would assume that oVirt [runs](https://github.com/oVirt/ovirt-engine/blob/936a8fcd09a4047c03b6cb5bf77709e0cba83537/backend/manager/modules/vdsbroker/src/main/java/org/ovirt/engine/core/vdsbroker/builder/vminfo/LibvirtVmXmlBuilder.java#L1619-L1620) VM with such a port?

**Reply by icsy7867** [reply]
Yeah I believe so.  So poking around:

<img width="570" alt="image" src="https://user-images.githubusercontent.com/3788162/206473407-f4487068-9da3-4c2a-9062-c60a550983f2.png">

It looks liek the file exists.  However
```
systemctl status qemu*
```
returns nothing

**Reply by tormath1** [reply]
What's the output of `cat /etc/os-release` ?

**Reply by icsy7867** [reply]
> What's the output of `cat /etc/os-release` ?

```
core@flatcar-master /dev/virtio-ports $ cat /etc/os-release
NAME="Flatcar Container Linux by Kinvolk"
ID=flatcar
ID_LIKE=coreos
VERSION=3374.2.0
VERSION_ID=3374.2.0
BUILD_ID=2022-11-15-1845
SYSEXT_LEVEL=1.0
PRETTY_NAME="Flatcar Container Linux by Kinvolk 3374.2.0 (Oklo)"
ANSI_COLOR="38;5;75"
HOME_URL="https://flatcar.org/"
BUG_REPORT_URL="https://issues.flatcar.org"
FLATCAR_BOARD="amd64-usr"
CPE_NAME="cpe:2.3:o:flatcar-linux:flatcar_linux:3374.2.0:*:*:*:*:*:*:*"
```

**Reply by tormath1** [reply]
Yeah, so as mentioned: this feature is available from release 3402, so everything < 3402 (like: 3374.2.0) won't have the qemu-guest-agent installed.
Can you try it with a beta image like this one: https://www.flatcar.org/releases#release-3417.1.0? Beta images are considered stable enough to run a few of them in your workload.

**Comment by icsy7867** [comment]
I see.  I didn't realize this was only available in the beta channel.  I'll give it a shot!
